src/lib.rs

```diff
@@ -311,14 +311,43 @@ pub unsafe extern "C" fn wgpuSurfaceGetPreferredFormat(
     surface: id::SurfaceId,
     adapter: id::AdapterId,
 ) -> native::WGPUTextureFormat {
-    let preferred_format = match wgc::gfx_select!(adapter => GLOBAL.surface_get_preferred_format(surface, adapter))
+    let preferred_format = match wgc::gfx_select!(adapter => GLOBAL.surface_get_supported_formats(surface, adapter))
     {
-        Ok(format) => conv::to_native_texture_format(format),
+        Ok(formats) => conv::to_native_texture_format(
+            *formats
+                .first()
+                .expect("Could not get preferred swap chain format"),
+        ),
         Err(err) => panic!("Could not get preferred swap chain format: {}", err),
     };
     return preferred_format;
 }
 
+#[no_mangle]
+pub unsafe extern "C" fn wgpuSurfaceGetSupportedFormats(
+    surface: id::SurfaceId,
+    adapter: id::AdapterId,
+    count: Option<&mut usize>,
+) -> *const native::WGPUTextureFormat {
```

Comment:
We can't actually return an owned allocation here because if we are running in a DLL (on windows) we actually have a different heap than the main process so them calling free would be invalid.

We have a couple options here:
- Force the user to allocate the memory and having them pass in a pointer and a length. The question becomes how to communicate to the user how much data to allocate. We could suggest "give us space for 8 formats" which is likely always enough. Otherwise we would need a function call to give them a length.
- Create a wgpuSurfaceSupportedFormatDelete function where we take back the pointer and delete it in our DLL. We'd need to be careful about this though (for example, for vectors length must == capacity when converting to a single length and pointer.

Reply:
Another option that we have used elsewhere in this repo, IIRC:
* Maintain ownership but never delete it. I.e. keep it a global store.

Reply:
added `wgpuFree`; it can be used to free allocations from rust side.

Reply:
well msvc doesn't like this macro
```c
#define WGPU_FREE(ptr, count) wgpuFree(ptr, count * sizeof(*ptr), _Alignof(typeof(*ptr)))

// usage
WGPU_FREE(usage, 1);
```

because it doesn't support expressions in `_Alignof` & also doesn't support `typeof` operator.
so macro will have to take in the type too, like
```c
#define WGPU_FREE(type, ptr, count) wgpuFree(ptr, count * sizeof(type), _Alignof(type))

// usage
WGPU_FREE(char, usage, 1); // users will have to be careful not to pass the pointer of type.
```

Reply:
and here's how it looks without macro

```c
char *usage = wgpuGetResourceUsageString();
printf("%s\n", usage);
wgpuFree(usage, 1 * sizeof(char), _Alignof(char));
```

personally i would just avoid the macro

Reply:
Cool, sounds good to me.